### PR TITLE
[Bugfix][fast] Fix the get_num_blocks_touched logic

### PR DESCRIPTION
**Pull request by zachzzc**
FILL IN THE PR DESCRIPTION HERE

FIX #xxxx (*link existing issues this PR will resolve*)

In the existing logics, the number of touched blocks is not calculated correctly. 

1) If the prefix caching is enabled, and the blocks are swapped out, the left-over the caches will be in the evictor group if there are no references. In the meantime, the cache blocks will be evicted, for example, 2 out of 10 cached blocks. Let's say the remaining 8 blocks in the evictor are the last available blocks in the allocator, and the sequence wants to swapped in, the free blocks will count as 8. The touched blocks is calculated as 2, which is incorrect because the cache is not shared by any other blocks. In this case the correct touched blocks should be 10, and can't swap in.

2) In the regular case, if the lookahead slots are more than the free slots, it first need to occupy the block containing the free slots, then plus the remaining required blocks.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>
<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by zachzzc** [comment]
@Kaiyang-Chen @cadedaniel  pls review when you have time.

**Comment by cadedaniel** [comment]
Thanks for the PR. Can you help me understand the impact here of this bug? e.g. what kind of incorrect swapping behavior / corrupted output.

Can we also add a unit test for the changed logic if possible? the block manager test is very high level unfortunately.

cc @Kaiyang-Chen do you have time to review this change?

**Comment by zachzzc** [comment]
> Thanks for the PR. Can you help me understand the impact here of this bug? e.g. what kind of incorrect swapping behavior / corrupted output.
> 

Hi Cade. The original problem I see is that when enable prefix caching and use swap for the preemption, the `allocate_block_id` raised `NoFreeBlocksError` with my input traces. After investigation, it is because in the `_schedule_swapped` function, `self.block_manager.can_swap_in` returns OK due to the problem I mentioned in the description. 

> Can we also add a unit test for the changed logic if possible? the block manager test is very high level unfortunately.
> 
sure I can add smaller tests for the prefix_caching_block and native blocks

**Comment by cadedaniel** [comment]
lmk when it's ready for review @zachzzc

**Comment by zachzzc** [comment]
> lmk when it's ready for review @zachzzc

Yes it is ready for review

File: tests/core/block/test_naive_block.py [hunk 1]
```diff
@@ -100,3 +100,42 @@ def test_get_num_free_blocks(allocate_type: str, num_blocks: int,
         for i, block in enumerate(blocks):
             assert allocator.get_num_free_blocks() == i
             allocator.free(block)
+
+    @staticmethod
+    @pytest.mark.parametrize("num_blocks", [4])
+    @pytest.mark.parametrize("block_size", [8])
+    def test_get_num_blocks_touched(num_blocks, block_size):
```
Comment (cadedaniel) [inline tests/core/block/test_naive_block.py:107]: Can you add a docstring explaining test methodology (e.g. what the test tests and how it does it)

File: tests/core/block/test_prefix_caching_block.py [hunk 2]
```diff
@@ -315,6 +315,55 @@ def test_get_num_free_blocks(num_blocks: int, block_size: int, seed: int):
                                                        i)
             allocator.free(block)
 
+    @staticmethod
+    @pytest.mark.parametrize("num_blocks", [4])
+    @pytest.mark.parametrize("block_size", [8])
+    def test_get_num_blocks_touched(num_blocks, block_size):
```
Comment (cadedaniel) [inline tests/core/block/test_prefix_caching_block.py:321]: Can you add a docstring explaining test methodology (e.g. what the test tests and how it does it)

**Review by cadedaniel — COMMENTED** [review]
Looks good, small comments

File: tests/core/block/test_naive_block.py (hunk 1 above)
Comment (zachzzc) [reply tests/core/block/test_naive_block.py:107]: done

File: tests/core/block/test_prefix_caching_block.py (hunk 2 above)
Comment (zachzzc) [reply tests/core/block/test_prefix_caching_block.py:321]: done.

File: tests/core/block/test_naive_block.py [hunk 3]
```diff
@@ -100,3 +100,45 @@ def test_get_num_free_blocks(allocate_type: str, num_blocks: int,
...
+                "mutable", allocator_src,
+                prev_block=src_blocks[-1],token_ids=[]
+            )
+        src_blocks.append(allocate_non_full_block())
+        src_blocks[-1].append_token_ids([0])
+
+        assert allocator_dst.get_num_blocks_touched(
+            src_blocks, num_lookahead_slots=1) == num_blocks
+        assert allocator_dst.get_num_blocks_touched(
+            src_blocks, num_lookahead_slots=block_size - 1) == num_blocks
```
Comment (Kaiyang-Chen) [inline tests/core/block/test_naive_block.py:142]: Can we make num_lookahead_slots a test parameter to follow the convention here?

File: tests/core/block/test_prefix_caching_block.py [hunk 4]
```diff
@@ -315,6 +315,59 @@ def test_get_num_free_blocks(num_blocks: int, block_size: int, seed: int):
...
```
Comment (Kaiyang-Chen) [inline tests/core/block/test_prefix_caching_block.py:321]: Can we merge two test_get_num_blocks_touched functions and control them by `enable_caching` parameter? Or at least make their name different?

**Review by Kaiyang-Chen — COMMENTED** [review]
Generally LGTM. Two small comments.

**Review by cadedaniel — APPROVED** [review]
LGTM, can merge once @Kaiyang-Chen 's comments are addressed

File: tests/core/block/test_prefix_caching_block.py (hunk 4 above)
Comment (zachzzc) [reply tests/core/block/test_prefix_caching_block.py]: Done, made the test name different.

File: tests/core/block/test_naive_block.py (hunk 3 above)
Comment (zachzzc) [reply tests/core/block/test_naive_block.py:142]: The three lookahead situations are block_size dependent here, I think it is more straightforward to leave it this way.

**Comment by cadedaniel** [comment]
@simon-mo can we get a force merge here

**Comment by simon-mo** [comment]
🫡

**Comment by cadedaniel** [comment]
thanks for the contribution @zachzzc !

**Pull request merged**